src/memfd.rs

```diff
@@ -215,6 +215,38 @@ impl Memfd {
         };
         Ok(r as u64)
     }
+
+    /// Set or clear the close-on-exec flag of the file descriptor.
+    ///
+    /// Note that you should normally create the memfd with the close-on-exec flag already set to prevent race conditions.
+    ///
+    /// It can be useful to disable the flag after creation to let the fd be inherited by a forked process.
+    /// Directly after forking the flag can be enabled again without race conditions,
+    /// because the forked process will have only one thread.
+    pub fn set_close_on_exec(&self, close_on_exec: bool) -> std::io::Result<()> {
```

Comment:
Is there any prior art in providing a method like this? e.g. `std::fs::File` does not provide a functionality like this.

Reply:
Internally, the standard library does enable the flag for created file descriptors in this manner as a fallback, in case the platform doesn't support creating the file descriptors with the flag already set.

I'm not aware of any public API that exposes it, besides the `filedesc` crate, but that's mine. Maybe other file descriptor wrapper libraries also expose it.

As I'm typing this, I'm thinking maybe the setter should be unsafe, since disabling the flag can cause leaked file descriptors. I'm not really sure if that can cause memory safety problems though.

Reply:
Leaked resources or their exhaustion is not considered to be a soundness problem, no.

Reply:
For reference, the `fd` crate has a [`set_flags`](https://docs.rs/fd/0.2.3/fd/fn.set_flags.html) function which can be used to do the same.

The `filedescriptor` crate doesn't seem to expose anything.

Reply:
Adding these methods would be a backwards-compatible change in either case. Given all the caveats setting cloexec has after creation, this method seems like an unnecessary footgun to me. As thus, I suggest we leave it out until there's a genuine use-case for this functionality. wdyt?

Reply:
Personally I don't need to disable the flag. I do have a use case for sharing the file descriptor, but we do that by passing it over a unix socket. I'm not really sure how well manually forking is supported in Rust anyway.

But to some extent the foot-gun is already there by having the option to create the file descriptor without the flag set in the first place. Should we remove that option too then?

I don't believe the standard library allows you to create any file descriptors without the flag set (apart from the race condition on platforms that can't set it at creation).

Reply:
I wouldn't be against removing the option to disable cloexec setting in the first place, but I also wouldn't mind _that_ one remaining too much.

The footgun I was referring to is the effective race you can get when toggling the flag on and off while there's forking action going on in other threads.

Reply:
It feels a bit incomplete to allow the flag to be manipulated, but only at creation. Personally I'd prefer an all or nothing here, although I'll concede to what you and @lucab prefer.

Reply:
I'm sorry, what I had in mind was something like a builder option, but I forgot that https://docs.rs/memfd/0.3.0/memfd/struct.MemfdOptions.html#method.close_on_exec actually already exists. That is already enough to cover my non-CLOEXEC usecase.

I agree that toggling it after the creation should not be encouraged, so we can avoid the `set_cloexec` method right now until somebody has an actual usecase for it.

I don't have a strong opinion on the `get_` portion of this (personally I don't have an usecase for that).

Reply:
Alright, I'll just remove the new functions then :)

Reply:
I did like the tests you added here in the previous iteration though, any chance that you could preserve them? The `get` logic could be moved into a test helper, I think.

Reply:
righty, will do :)

Reply:
added back tests to check to close-on-exec flag :)

Reply:
> The footgun I was referring to is the effective race you can get when toggling the flag on and off while there's forking action going on in other threads.

Btw, this doesn't have to be a footgun. You can disable the flag *after* the fork, before the exec. There is no race condition that way.

However, it's probably even nicer to `dup2` it to a lower, fixed number before to `exec`, so it's probably fine not to allow disabling the flag here.